s3driver.go

```diff
@@ -0,0 +1,216 @@
+package sftp
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+type S3 interface {
+	ListObjectsV2(input *s3.ListObjectsV2Input) (*s3.ListObjectsV2Output, error)
+	DeleteObject(input *s3.DeleteObjectInput) (*s3.DeleteObjectOutput, error)
+	CopyObject(input *s3.CopyObjectInput) (*s3.CopyObjectOutput, error)
+	PutObject(input *s3.PutObjectInput) (*s3.PutObjectOutput, error)
+	GetObject(input *s3.GetObjectInput) (*s3.GetObjectOutput, error)
+}
+
+type S3Driver struct {
+	s3       S3
+	bucket   string
+	homePath string
+}
+
+func (d S3Driver) Stat(path string) (os.FileInfo, error) {
+	localPath, err := translatePath(d.homePath, path)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := d.s3.ListObjectsV2(&s3.ListObjectsV2Input{
+		Bucket:  aws.String(d.bucket),
+		Prefix:  aws.String(localPath),
+		MaxKeys: aws.Int64(1),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.Contents == nil || *resp.KeyCount == 0 {
+		return nil, os.ErrNotExist
+	}
+
+	info := &fileInfo{
+		name: localPath,
+		mode: os.ModePerm,
+	}
+	if strings.HasSuffix(*resp.Contents[0].Key, "/") {
+		info.name = strings.TrimRight(info.name, "/")
+		info.mode = os.ModeDir
+	}
```

Comment:
Have you tested this with empty directories, non-empty directories, and non-existent directories?

Also, if the user doesn't request a directory, I believe this method could still return them a directory, e.g. if `path` is `/xyx` (a file), this will return `/xyz/` a directory if `xyz` exists as a directory. Is that how `Stat` is supposed to work?

In general I feel like this method has a bunch of edge cases worth testing on S3, especially with S3's strange treatment of directories.